game/theater/iadsnetwork/iadsnetwork.py

```diff
@@ -249,17 +250,21 @@ def initialize_network_from_range(self) -> None:
                 except IadsNetworkException:
                     # TGO does not participate to iads network
                     continue
-                # Find nearby Power or Connection
-                for nearby_go in self.ground_objects.values():
-                    if nearby_go == go:
-                        continue
-                    if (
-                        IadsRole.for_category(go.category)
-                        in [
-                            IadsRole.POWER_SOURCE,
-                            IadsRole.CONNECTION_NODE,
-                        ]
-                        and nearby_go.position.distance_to_point(go.position)
-                        <= node.group.iads_role.connection_range.meters
-                    ):
-                        node.add_connection_for_tgo(nearby_go)
+                self._make_advanced_connections(node)
+
+    def _is_friendly(self, node: IadsNetworkNode, tgo: TheaterGroundObject) -> bool:
+        node_friendly = node.group.ground_object.is_friendly(True)
+        tgo_friendly = tgo.is_friendly(True)
+        return node_friendly == tgo_friendly
+
+    def _make_advanced_connections(self, node: IadsNetworkNode) -> None:
+        tgo = node.group.ground_object
+        # Find nearby Power or Connection
+        for nearby_go in self.ground_objects.values():
+            iads_role = IadsRole.for_category(nearby_go.category)
+            if not iads_role.is_comms_or_power or nearby_go == tgo:
+                continue
+            dist = nearby_go.position.distance_to_point(tgo.position)
+            in_range = dist <= iads_role.connection_range.meters
+            if in_range and self._is_friendly(node, nearby_go):
```

Comment:
The last thing that was forgotten, for obvious reasons we ought to check if `nearby_go` is friendly...

Reply:
During intialization of the iads network we dont care about friendly or not. The network is the whole network of all connections independent from nodes being friendly or not. Otherwise you have to always calculate the connections again whenever a capture event hits the deck. So the check for is friendly was added to the representation logic (the Map and the `skynet_nodes` function which will generate the connections for the skynet plugin.
So is friendly is actually breaking the system.

Reply:
>Otherwise you have to always calculate the connections again whenever a capture event hits the deck.

Why is this a problem exactly? Especially if we're initializing by range I would say it's only logical...

But I might be missing something that relates to the skynet plugin, cause tbh I'm not sure how the magic works in DCS itself. I guess I should go take a look to see what skynet is expecting exactly because I'm starting to ask myself some questions. Like, does skynet actually expect the entire network to stay connected? Does skynet itself make the distinction between friend and foe?

Reply:
It depends on how want to approach this problem:
1) calculate the connections to all nearby objects without taking care if they are friendly and later check if the connection is "active"
2) Always recalculate the connections from one specific TGO whenever there are some changes to this tgo and only represent the current state.

Solution 2 has one problem: what happens when a nearby comm tower gets captured? Will the connection from the nearby SAM System to the comm tower be recreated if the sam system is under the hood of a different CP?

I personally do not care how to solve this. I've gone with approach one so that i dont have to iterate over -all- TGOs all the time something changes (every round with a capture or destroy). Whatever you choose keep this one in mind.

Reply:
Solution 2 is what this whole PR is about, including a fix for the problem you're pointing out. But I agree that iterating over all TGOs is not the way to go. I could build further on this and make container classes for comm towers and power stations so they have a list of nodes that connect to them, making it easier to access those nodes to update their connection to the comm tower or power station in question. Effectively I think this would be sort of a hybrid solution between 1 and 2 if you see where I'm going with this...

Reply:
Also, I'm not sure what you mean by checking if the connection is "active", as this doesn't consider whether or not the unit in question is friendly or not. Looking at the JS model, it just checks if both ends of the connection have units alive. (see lines 48-50)
https://github.com/dcs-liberation/dcs_liberation/blob/cc4237d0762cea396fe56134adf2bc64fa1682c4/game/server/iadsnetwork/models.py#L32-L57

Reply:
> Solution 2 has one problem: what happens when a nearby comm tower gets captured? Will the connection from the nearby SAM System to the comm tower be recreated if the sam system is under the hood of a different CP?

Very good question, this was definitely a problem so let me try to elaborate:

So a capture event occurs, all TGOs are depopulated so that guarantees all IADS nodes connected to the CP in question are "not participating". Comm towers and power stations aren't nodes though, so they don't get updated. That's a problem being fixed by #2285 & #2286 because the logic is missing. What you pointed out about not wanting to iterate over all TGOs whenever there's a change is something I see as necessary with the way things are currently designed. I recommend looking at `_update_iads_comms_and_power`, I'll open a conversation there and tag you.

Reply:
So a quick look at the readme for skynet iads [here](https://github.com/walder/Skynet-IADS#quick-start) suggests that multiple networks for both coalitions are possible:
> A Skynet IADS is a complete operational network. You can have multiple Skynet IADS instances per coalition in a DCS mission. A simple setup would be one IADS for the blue side and one IADS for the red side.

From the simple setup I would assume connections between blue and red are not allowed, though I'm not sure because it's also possible that skynet simply ignores connections between nodes that aren't from the same coalition. It's something I'm trying to figure out by going through the code but no luck so far.

Either way, I still haven't understood why connecting to friendly TGOs exclusively is a breaking change, perhaps you can clear that up for me?

Reply:
You already quoted the line which checks if the connection should be plotted or not. Currently the logic to tell if a connection is on red or blue side is all happening outside the IADSNetwork. Check SkynetNode.from_group there is the property player which defines if the Node should be added to the RED IADS or BLUE IADS in terms of the skynet relevant code. On the other side for the map, well you quoted it ;)
For me it is completly irrelevant if you go with solution 1 or solution 2 as long as you take care of the problem i mentioned above. So to be clear: there is no breaking change which would prevent you from going with solution 2. Totally up to you ;)

Reply:
You can find the skynet lua data in the luagenerator.py and also take a look into a generated mission. There is a trigger i think it is called Liberation Data or similar. The liberation tailored lua code (skynetiads-config.lua) for the skynet plugin just goes through this data table

Reply:
> You already quoted the line which checks if the connection should be plotted or not. Currently the logic to tell if a connection is on red or blue side is all happening outside the IADSNetwork.

Right, but the problem then remains w.r.t. existing connections when comm towers and/or power stations flip sides after a capture. What I could do is keep it connected "under the hood" and only send a "delete event" to the UI. Though I still think the "hybrid solution" would be more elegant and readable, but then I'd prefer having the PRs merged before I refactor it. Either way, I sort of depend on the answer to my next question below, because if the skynet plugin ignores connections between RED and BLUE, then I shouldn't be deleting connections "under the hood", otherwise I think we'd be better off "sanitizing" the connections.

> Check SkynetNode.from_group there is the property player which defines if the Node should be added to the RED IADS or BLUE IADS in terms of the skynet relevant code.

But that's referring to the node, not its connections as far as I can tell. So theoretically I think RED IADS could be connected to BLUE IADS, and that just seems very contradicting to me. It's one of the things I'm still trying to clear up, i.e. does the skynet plugin itself ignores those connections?

Reply:
So I think I have an answer. I used Golan Heights to see how skynet behaves since it's an example of what I described, i.e. in the campaign file I find connections between blue and red, though they don't seem to cause a problem so the whole is_friendly thing is indeed redundant. I'll adjust this PR, as well as #2285, #2286 and #2287, knowing what I know now I believe I have more redundant stuff that needs to be cleaned up.

Reply:
Also this would mean to rewrite the whole lua script as well. I prefer the friend and foe seperation / blue & red so that we can stick to the logic skynet wants from us. This will also match your preference with the "solution 2" thing as you said you want to always recreate the network with the real connections. So this would match.
Maybe i missed something but so far this should not be an issue at all

Reply:
> I never tested adding hostile nodes to the iads in skynet but i would assume it does have some side effects. I would suggest not doing so and only add the friendly nodes.

You do realize `golan_heights_lite.yaml` connects `Golan North-52` to `IADS_CN3` and `IADS_CN5`. To elaborate, GN52 is an enemy EWR unit while one of the connections nodes (not sure if it's CN3 or CN5 but I think it's CN3) is friendly. Was this intentional?

Reply:
The iads config is used for the iadsnetwork initialization only. As i said the initial approach was to have an independent network and only during runtime (skynet_nodes function) we calculate if a connection should be handed over to the skynet plugin by checking if primary and connection node are friendly.

Reply:
Riiiiight, don't know how I missed that one. I think everything's clear now, so that means the check for is_friendly is redundant in a way, but most definitely not a breaking change. In a way perhaps even favorable because intuitively it only seems logical to connect exclusively with friendly nodes.

Reply:
But to resolve this issue, where do we stand on the initialization by range? Because all this makes me conclude that during initialization by range we can filter for friendly connections without causing any problem, right?